pkg/argocd/update.go

```diff
@@ -408,27 +411,53 @@ func marshalParamsOverride(app *v1alpha1.Application, originalData []byte) ([]by
 		mergeKustomizeOverride(&params, &newParams)
 		override, err = yaml.Marshal(params)
 	case ApplicationTypeHelm:
-		if app.Spec.Source.Helm == nil {
+		if appSource.Helm == nil {
 			return []byte{}, nil
 		}
-		var params helmOverride
-		newParams := helmOverride{
-			Helm: helmParameters{
-				Parameters: app.Spec.Source.Helm.Parameters,
-			},
-		}
 
-		if len(originalData) == 0 {
-			override, err = yaml.Marshal(newParams)
-			break
-		}
-		err = yaml.Unmarshal(originalData, &params)
-		if err != nil {
-			override, err = yaml.Marshal(newParams)
-			break
+		if strings.HasPrefix(app.Annotations[common.WriteBackTargetAnnotation], common.HelmPrefix) {
+			images := GetImagesFromApplication(app)
+
+			for _, c := range images {
+				// Build string with YAML format to merge with originalData values
+				helmValues := fmt.Sprintf("%s: %s\n%s: %s",
+					app.Annotations[fmt.Sprintf(common.HelmParamImageNameAnnotation, c.ImageName)],
+					getHelmParam(appSource.Helm.Parameters, app.Annotations[fmt.Sprintf(common.HelmParamImageNameAnnotation, c.ImageName)]).Value,
```

Comment:
Also, while testing the image I found out that, if you set `argocd-image-updater.argoproj.io/write-back-target: helmvalues` (or similar, with path) annotation but you don't set `helm.image-name` and `helm.image-tag` annotation, it will not take the defaults, the code crashes. This should be fixed.

Reply:
Thanks for the feedback. That's true, I did not check on because I didn't expect to use the helmvalues feature without the image name and image tag.
I'll fix this asap.

Again, thank you :)

Reply:
I'll add a function to check if those annotations are in place and raise an error otherwise. Hopefully I'll update the code this week. @mihaelabalas84 , thank you for your help :)

Reply:
I made some fixes for this bug. Instead of building them manually, I just used a function that was already there to retrieve those values and I did not see it before.

I pushed those changes a few minutes ago. Also there are some new tests to validate that behaviour.

Please, let me know if you find some more issues.
@mihaelabalas84 , again, thank you for your help :)

Reply:
I'll mark this conversation as resolved, please let me know if there is something about the changes to solve this issue that may be wrong.